### PR TITLE
Update docs/transform.ipynb based on new transformer framework.

### PR DESCRIPTION
**Pull request by tanujkhattar**
Fixes https://github.com/quantumlib/Cirq/issues/4960

This notebook provides a high level overview of all the newly added features to Cirq's transformer infrastructure. We can discuss whether we need more details docs on each of the specific topics -- I think a high level overview with links to reference pages of individual methods (i.e. the current PR) should be enough to close https://github.com/quantumlib/Cirq/issues/4960, but I'm curious to hear what others think.

### CODE REVIEW

File: docs/transform.ipynb
Comment (MichaelBroughton) [inline docs/transform.ipynb:178]: Nit: generally philosphy is to instruct the reader and not sell/editorialize where possible.

File: docs/transform.ipynb
Comment (MichaelBroughton) [inline docs/transform.ipynb:217]: Maybe add a code cell to further clarify this difference via some output comparison ?

File: docs/transform.ipynb
Comment (MichaelBroughton) [inline docs/transform.ipynb:426]: Personally I would delete the laundry list here and just link out to the `cirq.transformers` module so a motivated user could sift through the latest features there. Plus as maintainers we don't have to remember to keep this up to date.

File: docs/transform.ipynb
Comment (MichaelBroughton) [inline docs/transform.ipynb:445]: It feels like there might be a bit too much information in this text here. Could we cut this down to introduce `optimize_for_target_gateset` maybe demo a built-in gateset and end on some hook to encourage the reader to check out other CompilationTargetGatesets to see how they work ?

File: docs/transform.ipynb
Comment (MichaelBroughton) [inline docs/transform.ipynb:470]: Nit: break up into multiple cells and walk through what's happening clearly.

File: docs/transform.ipynb
Comment (MichaelBroughton) [inline docs/transform.ipynb:496]: Usually we want to end on some kind of "Now you go and build on this for yourself". Could we rework this into something like: "Try using transformers to optimize your circuits for your compilation needs" or "try adapting the above transformers to work with SycamoreGates" or something ?

**Review by MichaelBroughton — COMMENTED** [review]
Good first pass. Left a few high level comments, my biggest concern is that things are a little information dense. We don't need to explain lots of details in the tutorials, but give a nice walkthrough of how things work from a basic to moderately advanced use case and then springboard people into the right parts of the API docs if they want to become experts. IMO I think this would read a lot better if we moved the `Built in transformers in Cirq` section up a little higher and simplified the introduction so that users could learn to be comfortable with the built in transformers, then they could gradually upgrade to a standalone custom function, standalone custom class and then CompilationTargetGateset and then done. Right now it seems a little out of order and a little information dense.

**Comment by tanujkhattar** [comment]
@MichaelBroughton Restructured the entire notebook as per the feedback. We now introduce built-in transformers first, show examples of how to use transformer context and then move on to details about the API / decorator and give examples on how users can create their own transformers. 

Please take another look.

File: docs/transform.ipynb
Comment (tanujkhattar) [reply docs/transform.ipynb:96]: Removed the type hints.

File: docs/transform.ipynb
Comment (tanujkhattar) [reply docs/transform.ipynb:122]: Restructured. PTAL.

File: docs/transform.ipynb
Comment (tanujkhattar) [reply docs/transform.ipynb:771]: removed type hints.

File: docs/transform.ipynb
Comment (tanujkhattar) [reply docs/transform.ipynb]: My motivation to keep this list as it is is that `cirq.transformers` reference page on the docs site has too many functions (eg: all functions from transformers.analytical_decompositions, transformers.heuristic_decompositions etc.) and it's hard to look at the reference page and guess which of these are transformers. 

Can we modify the reference page to have a classification of functions into transformers / transformer primitives / analytical decompositions etc?

Till we don't have that, I think it's worth keeping this list here.

File: docs/transform.ipynb
Comment (tanujkhattar) [reply docs/transform.ipynb]: Reduced the information content and restructured. PTAL.

File: docs/transform.ipynb
Comment (tanujkhattar) [reply docs/transform.ipynb]: Added an ending line with link to api reference docs.

File: docs/transform.ipynb
Comment (tanujkhattar) [reply docs/transform.ipynb]: done

File: docs/transform.ipynb
Comment (tanujkhattar) [reply docs/transform.ipynb]: Restructured, this is moot now.

File: docs/transform.ipynb
Comment (tanujkhattar) [reply docs/transform.ipynb:750]: done

File: docs/transform.ipynb
Comment (tanujkhattar) [reply docs/transform.ipynb]: reworded.

File: docs/transform.ipynb
Comment (MichaelBroughton) [inline docs/transform.ipynb:135]: Nit: we show. Second person you might want to do something like "Below, you can implement a transformer pipeline called `optimize_circuit`".

File: docs/transform.ipynb
Comment (MichaelBroughton) [inline docs/transform.ipynb:209]: Nit: second person

File: docs/transform.ipynb
Comment (MichaelBroughton) [inline docs/transform.ipynb:669]: Nit: second person.

File: docs/transform.ipynb
Comment (MichaelBroughton) [inline docs/transform.ipynb:338]: Nit: second person.

**Review by MichaelBroughton — APPROVED** [review]
Looks awesome. I think this flows really nicely now. A few more nits and then one final step: Generally we want to ship tutorials with no outputs in the cells (and let the docs jobs populate them for us). In our docs infra a tutorial that has pre-computed notebook cells will not get run on the docs infra and instead just be rendered directly to the site. For stability/reliability reasons its often better to leave notebooks without any output if we can.

After that feel free to merge.

**Pull request merged**
